src/librbd/crypto/CryptoInterface.h

```diff
@@ -15,7 +17,102 @@ class CryptoInterface : public RefCountedObject {
 public:
   virtual int encrypt(ceph::bufferlist* data, uint64_t image_offset) = 0;
   virtual int decrypt(ceph::bufferlist* data, uint64_t image_offset) = 0;
+  virtual uint64_t get_block_size() const = 0;
 
+  inline std::pair<uint64_t, uint64_t> get_pre_and_post_align(
+          uint64_t off, uint64_t len) {
+    if (len == 0) {
+      return std::make_pair(0, 0);
+    }
+    auto block_size = get_block_size();
+    return std::make_pair(p2phase(off, block_size),
+                          p2nphase(off + len, block_size));
+  }
+
+  inline std::pair<uint64_t, uint64_t> align(uint64_t off, uint64_t len) {
+    auto aligns = get_pre_and_post_align(off, len);
+    return std::make_pair(off - aligns.first,
+                          len + aligns.first + aligns.second);
+  }
+
+  inline bool is_aligned(uint64_t off, uint64_t len) {
+    auto aligns = get_pre_and_post_align(off, len);
+    return aligns.first == 0 && aligns.second == 0;
+  }
+
+  inline bool is_aligned(const io::ReadExtents& extents) {
+    for (const auto& extent: extents) {
+      if (!is_aligned(extent.offset, extent.length)) {
+        return false;
+      }
+    }
+    return true;
+  }
+
+  inline void align_extents(const io::ReadExtents& extents,
+                            io::ReadExtents* aligned_extents) {
+    for (const auto& extent: extents) {
+      auto aligned = align(extent.offset, extent.length);
+      aligned_extents->emplace_back(aligned.first, aligned.second);
+    }
+  }
+
+  inline int decrypt_aligned_extent(io::ReadExtent& extent,
+                                    uint64_t image_offset) {
+    if (extent.length == 0) {
+      return 0;
+    }
+
+    if (extent.extent_map.empty()) {
+      extent.extent_map.emplace_back(extent.offset, extent.bl.length());
+    }
+
+    ceph::bufferlist result_bl;
+    io::Extents result_extent_map;
+
+    ceph::bufferlist curr_block_bl;
+    auto curr_offset = extent.offset;
+    auto curr_block_start_offset = curr_offset;
+    auto curr_block_end_offset = curr_offset;
+
+    // this will add a final loop iteration for decrypting the last extent
+    extent.extent_map.emplace_back(
+            extent.offset + extent.length + get_block_size(), 0);
+
+    for (auto [off, len]: extent.extent_map) {
+      auto [aligned_off, aligned_len] = align(off, len);
+      if (aligned_off > curr_block_end_offset) {
+        curr_block_bl.append_zero(curr_block_end_offset - curr_offset);
```

Comment:
Technically this shouldn't be possible, right? If the encryption cipher block is 16 bytes, I would suspect that inserting a zero would change the remainder of the bytes, right? The only way this could happen is if a discard was issued, but then we would want to ensure that discards are aligned to the 4KiB block size to prevent random missing bytes.

Then if you don't need to worry about inserting zeroes, I think the remainder of this code could be simplified since the extent-map and read extents should be properly aligned with only the possibility of having a block-size multiple hole in the extent-map.

Reply:
In our mail thread you said:
However, I would still ensure that each sparse extent you get back
from the OSD is block-aligned (or just assume any missing sections are
zeroed).

So my assumption was that maybe the OSD can (potentially in a future version) decide to optimize out random zeros that were part of a ciphertext, and that's why we want to pad with zeros.
Do you want me to assume that each extent in the extent_map is crypto-aligned?
In that case I can just iterate the extents and decrypt them.

Reply:
Now I recall another reason that I thought of that supports keeping the more complicated code:
This code may be re-used (replacing decrypt with encrypt) when implementing prepare_copyup where I have to encrypt a SparseBufferList, where extents need not be crypto aligned.